api/cpp/include/slint_platform.h

```diff
@@ -0,0 +1,233 @@
+    {
+    }
+    virtual ~WindowAdapter() = default;
+    WindowAdapter(const WindowAdapter &) = delete;
+    WindowAdapter &operator=(const WindowAdapter &) = delete;
+
+    /// Return a reference to the renderer that can be used to do the rendering.
+    const R &renderer() const { return m_renderer; }
+
+    /// This function is called by Slint when the slint window is shown.
+    ///
+    /// Re-implement this function to forward the call to show to the native window
+    virtual void show() const { }
+    /// This function is called by Slint when the slint window is hidden.
+    ///
+    /// Re-implement this function to forward the call to hide to the native window
+    virtual void hide() const { }
+
+    /// This function is called when Slint detects that the window need to be repainted.
+    ///
+    /// Reimplement this function to forward the call to the window manager.
+    ///
+    /// You should not render the window in the implementation of this call. Instead you should
+    /// do that in the next iteration of the event loop, or in a callback from the window manager.
+    virtual void request_redraw() const { }
+
+    /// Return the slint::Window associated with this window.
+    ///
+    /// Note that this function can only be called if the window was initialized, which is only
+    /// the case after it has been returned from a call to Platform::create_window_adaptor
+    Window &window()
+    {
+        if (!was_initialized)
+            std::abort();
+        // This works because cbindgen_private::WindowAdapterRcOpaque and Window have the same
+        // layout
+        return *reinterpret_cast<Window *>(&self);
+    }
+
+    /// Send a pointer event to this window
+    // Note: in rust, this is on the Window. FIXME: use a public event type
+    void dispatch_pointer_event(const cbindgen_private::MouseEvent &event)
+    {
+        private_api::assert_main_thread();
+        if (was_initialized) {
+            cbindgen_private::slint_windowrc_dispatch_pointer_event(&self, event);
+        }
+    }
+
+    /// Send a key event to this window
+    // Note: in rust, this is on the Window. FIXME: use a public event type
+    void dispatch_key_event(const cbindgen_private::KeyInputEvent &event)
+    {
+        private_api::assert_main_thread();
+        if (was_initialized) {
+            cbindgen_private::slint_windowrc_dispatch_key_event(&self, event);
+        }
+    }
+
+    /// Returns true if the window is currently animating
+    bool has_active_animations() const
+    {
+        return cbindgen_private::slint_windowrc_has_active_animations(&self);
+    }
+};
+
+/// The platform is acting like a factory to create a WindowAdapter<R>
+///
+/// Platform::register_platform() need to be called before any other Slint handle
+/// are created, and if it is called, it will use the WindowAdapter provided by the
+/// create_window_adaptor function.
+template<Renderer R>
```

Comment:
Ah, so you decided against a base class but keep the renderer as a parameter of the Platform?

I do think it's a little weird that the platform is parameterised with the renderer.

Or planned for follow-up?

Reply:
planned follow up